### ISSUE TITLE
Better documentation for making custom generators

### ISSUE THREAD

**Issue by Timmmm** [emil-e/rapidcheck#178]
I really want to like RapidCheck but at the moment it is really really hard to see how to do custom generators. For example I want to generate a `vector<int>` whose product is a sane random number. I have prototyped this in Python:

```
def gen(size):
    len = (size+19) // 20
    if len == 0:
        return []

    rand = [random.uniform(0, 4) for x in range(0, len)]
    rand_sum = reduce(add, rand, 0)

    product = random.uniform(0, size * 100)

    c = (math.log(product) - rand_sum) / len

    return [int(math.exp(r + c)) for r in rand]
```

It works fine, but I am at a total loss as to how to get this into RapidCheck. There aren't any examples that don't make use of existing uniform distributions like `inRange()` or are just setting simple fields of a class. I managed to get a little way but the code is so full of templates and templates of templates that it is very difficult to follow. Especially as there aren't many comments. For example, I looked at copying the container generator which uses this function (which has no comments):

```
template <typename T, typename Shrink>
Shrinkable<Decay<T>> shrinkRecur(T &&value, const Shrink &shrinkf) {
  return shrinkable::shrink(fn::constant(std::forward<T>(value)),
                            [=](Decay<T> &&x) {
                              return seq::map(shrinkf(std::move(x)),
                                              [=](Decay<T> &&xshrink) {
                                                return shrinkable::shrinkRecur(
                                                    std::move(xshrink),
                                                    shrinkf);
                                              });
                            });
}
```

I mean. Like... *one* comment would be nice. What does this do? To use it I'd basically have to learn how the entire code-base works.

Some other basic things the documentation is lacking:

* There's this "size" parameter, but what exactly is it? If I generate an `int` with size 50, what distribution is that `int` drawn from?
* How do I control the distribution of numbers?
* Similarly for `vectors` and other containers. How does the "size" affect their length? How does it affect their contents? Is it passed through to each element?
* How is `*rc::gen::inRange(0, 100)` different from just using `rand() % 100`?
* How does the shrinking work exactly? Presumably the return value of each generator is remembered somehow?

I hope this doesn't seem like too much negative criticism. I've used RapidCheck successfully at a previous job and when it works it is really useful. And it is free! It's just that the code is very template-spaghetti and hard to follow, so it's a herculean task to venture beyond the limited examples.

**Comment by emil-e** [comment]
The documentation on how to implement custom generators is, admittedly, lacking. Since I do not get paid for working on RapidCheck, documentation on more advanced topics has taken a backseat.

The effect of size is loosely specified on purpose but is designed to be appropriate for many common use cases. This approach is directly taken from the original QuickCheck implementation. If you need to explicitly control what it means, you should probably implement your own custom generator where you can define that for yourself or use a generator where the specific things you're after are not affected by size. Explicit control over distributions is not provided by RapidCheck at this time, you'll need to create your own generator for this.

Using `rand() % 100` does not allow RapidCheck to control the return value and thus it cannot implement shrinking. `*rc::gen::inRange(0, 100)` in this case would would generate a number in the range of [0, 100) and would use only (a lower) part of the range for small sizes but grow that to the entire range once you get up to `kNominalSize`.

Shrinking, in general terms, works by `Gen<T>::operator()` returning a `Shrinkable<T>`. A `Shrinkable<T>` is an object representing the specific value (returned by `.value()`) and a `Seq` of its possible shrink in order of ascending "size". These shrinks are also `Shrinkable`s which means that `Shrinkable` is actually a value together with the entire tree of possible shrinks for that value. This tree is possibly infinitem but since it is generated lazily (`Seq`s are lazy sequences and `Shrinkable` itself is lazy), that's okay.

To create a `Shrinkable`, what you usually do is to use the function that you mention above, `shrinkRecur`. This function takes a "root" value and a function that given a value returns a `Seq` of possible shrinks. The returned `Shrinkable` will represent the tree of shrinks you would get if you applied this function recursively. You can further manipulate this by using combinators such as `shrink::map`, `shrink::mapcat` and others. You can find them in the `.h` files (the `.hpp` files have the definition but no docs, as you noticed) in `include/shrinkable` and some further utils in `include/shrink`. They should all have comments but I can't promise you find them helpful.
